### PR TITLE
[Segmentation] Added generalized dice score metric

### LINKED ISSUE #1089: Generalized Dice Score
## 🚀 Feature

Generalized Dice Score (= 1 - [Generalized Dice Loss](https://arxiv.org/abs/1707.03237)), to evaluate overall segmentation performance in multiclass scenarios.

### Motivation

Nice complement to Dice Score. 

### Pitch

Addition of new class and function for generalized dice score computation.

### Additional context

Will make a PR for this right away

### PR DESCRIPTION
**Pull request by jlcsilva**
## What does this PR do?

Fixes #1089 . Didn't manage to make it work for multidimensional multilabel scenarios. However, that shouldn't a very common use case.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
I beilve that this shall be automatically resolved in cmd by `git merge upstream/master`

**Comment by Borda** [comment]
@jlcsilva shall be resolved now, pls check if I did not make a mistake... :rabbit:

**Comment by jlcsilva** [comment]
> @jlcsilva shall be resolved now, pls check if I did not make a mistake... 🐰

Seems ok to me. 

Also, I know why the tests are failing, but I'm unsure about what to do. In the file tests/unittests/classification/inputs.py
I changed the data type of some input tensors, such as _input_binary, from int to float, to match the description [here](https://torchmetrics.readthedocs.io/en/latest/pages/classification.html#input-types). Since the tests are failing, I'm not sure what's wrong, the documentation or the code. Any thoughts from someone with a better overall view of the repo?

**Comment by SkafteNicki** [comment]
Hi @jlcsilva, when looking at the changes in the PR there also seems to be some changes to other metrics. Any particular reason for this?

**Comment by jlcsilva** [comment]
> Hi @jlcsilva, when looking at the changes in the PR there also seems to be some changes to other metrics. Any particular reason for this?

In the StatScore class I added the None option for the reduce argument, which I needed for my generalized dice score implementation, and changed its documentation and that of its subclasses accordingly. Additionally, I corrected some typos and added some missing information to these classes' documentation. 

Finally, I changes the data type of some test inputs to match that in the general input type documentation page, and added new multidimensional input tensors. To me, this seems to be what is making the code fail the tests. Currently, I'm out of office, but I'll check it as soon as I go back to work

File: src/torchmetrics/classification/generalized_dice.py
Comment (Borda) [inline src/torchmetrics/classification/generalized_dice.py:75]:
```suggestion
            Determines whether the input is ``multidim`` or not.
```

File: src/torchmetrics/classification/generalized_dice.py
Comment (Borda) [inline src/torchmetrics/classification/generalized_dice.py:49]:
```suggestion
            numerator is also 0, and 0 otherwise.
```

File: src/torchmetrics/functional/classification/generalized_dice.py
Comment (Borda) [inline src/torchmetrics/functional/classification/generalized_dice.py:67]: this seems to be void

File: src/torchmetrics/functional/classification/generalized_dice.py
Comment (Borda) [inline src/torchmetrics/functional/classification/generalized_dice.py:148]:
```suggestion
            numerator is also 0, and 0 otherwise.
```

File: src/torchmetrics/functional/classification/generalized_dice.py
Comment (Borda) [inline src/torchmetrics/functional/classification/generalized_dice.py:173]:
```suggestion
        multiclass:
            Determines whether the input is multiclass (if True) or multilabel (if False).

        multidim:
            Determines whether the input is multidim or not.
```
duplicated from func signature

File: src/torchmetrics/classification/generalized_dice.py
Comment (Borda) [inline src/torchmetrics/classification/generalized_dice.py:72]:
```suggestion
            Determines whether the input is multiclass (if True) or multilabel (if False).
```

File: tests/classification/test_generalized_dice_score.py
Comment (Borda) [inline tests/classification/test_generalized_dice_score.py:51]: I do not see any import or use of scikit-learn so why do you name it this way?

File: src/torchmetrics/classification/generalized_dice.py
Comment (Borda) [inline src/torchmetrics/classification/generalized_dice.py:75]:
```suggestion
            Determines whether the input is multidim or not.
```

**Review by SkafteNicki — CHANGES_REQUESTED** [review]
Hi @jlcsilva,
Sorry for the long delay. We are currently doing a major refactor of the classification package #1001 and are therefore putting this PR on hold for now. When the refactor is done we will mix merge conflicts and afterwards focus on getting your awesome work merged :)

**Comment by jlcsilva** [comment]
Hey @SkafteNicki,
No problem! I've also been away, so wasn't able to help. I noticed the changes when I came back and pulled the repo again. I'm actually a bit busy today and next week, but let me know if there is something I can help with :)

**Comment by Borda** [comment]
@jlcsilva @SkafteNicki, mind checking the collisions after the refactor was merged... :)

**Comment by Borda** [comment]
@jlcsilva @stancld, mind checking the collisions after the refactor was merged... :otter:

**Comment by stancld** [comment]
@Borda I'll do rebase now.

**Comment by stancld** [comment]
ℹ️ It will be needed to add test cases after refactoring tests as a part of the classification refactor.

**Review by japui-coder — CHANGES_REQUESTED** [review]
tests/integrations/test_lightning.py

**Comment by SkafteNicki** [comment]
Going to take over this PR now so we hopefully can land it within the next week :]

**Comment by Borda** [comment]
> Going to take over this PR now so we hopefully can land it within the next week :]

hi @SkafteNicki how is it going here? :)

**Comment by Borda** [comment]
@justusschock, how much need to be done? :chipmunk:

File: src/torchmetrics/classification/dice.py
Comment (Borda) [inline src/torchmetrics/classification/dice.py:29]:
```suggestion
    .. deprecated:: v0.12
```

File: src/torchmetrics/classification/dice.py
Comment (Borda) [inline src/torchmetrics/classification/dice.py:30]:
```suggestion
        The `Dice` module was deprecated in v0.12 and will be removed in v0.13. Use `F1Score` module instead which
```

File: src/torchmetrics/classification/dice.py
Comment (Borda) [inline src/torchmetrics/classification/dice.py:153]:
```suggestion
            "The `dice` function was deprecated in v0.12 and will be removed in v0.13. Use `f1score` function instead"
```

File: src/torchmetrics/classification/generalized_dice.py
Comment (Borda) [inline src/torchmetrics/classification/generalized_dice.py:212]:
```suggestion
        >>> from torch import tensor
        >>> from torchmetrics import GeneralizedDiceScore
        >>> preds = tensor([2, 0, 2, 1])
        >>> target = tensor([1, 1, 2, 0])
```

File: src/torchmetrics/functional/classification/generalized_dice.py
Comment (Borda) [inline src/torchmetrics/functional/classification/generalized_dice.py:218]:
```suggestion
```
protected does not need examples

File: src/torchmetrics/functional/classification/generalized_dice.py
Comment (Borda) [inline src/torchmetrics/functional/classification/generalized_dice.py:368]:
```suggestion
        >>> from torch import tensor
        >>> from torchmetrics.functional import generalized_dice_score
        >>> preds = tensor([2, 0, 2, 1])
        >>> target = tensor([1, 1, 2, 0])
```

**Review by Borda — APPROVED** [review]
except https://github.com/Lightning-AI/metrics/pull/1090#discussion_r909536311it looks good to me

**Comment by Borda** [comment]
@jlcsilva, any chance you could work on this one, in particular, add some tests... :flags:

**Comment by Borda** [comment]
seems we hit some internal Gh issue, will be restarted later...

**Review by SkafteNicki — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]
So happy to see it landing 🤩

**Pull request merged**
